fix: add modifying nested struct diagnostic

TimothyMakkison opened the pull request:
# Add diagnostic for modifying a temporary value type

## Description
Added check for if a nested value type is being modified.

Probably not a common issue so I went with a diagnostic.
I considered nested chain assignment or a `with` expression ie `target.NestedValue = target.NestedValue with { StringValue = src.StringValue };` (can repeatedly chain)


Fixes #516

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/ObjectMemberMappingBodyBuilder.cs:
```diff
@@ -135,6 +135,27 @@ PropertyMappingConfiguration config
+        // cannot assign to intermediate value type
+        if (targetMemberPath.Path.Count > 1)
+        {
+            var member = targetMemberPath.Path.Reverse().Skip(1).First();
```
This only checks the second last item, but couldn't be any in the path except for the last item be a problem?
Otherwise this could be simplified to `var member = targetMemberPath.Path[^2];` / `var member = targetMemberPath.Path[targetMemberPath.Path.Count - 2];` (Index doesn't work due to net standard 2.0).

latonz commented on test/Riok.Mapperly.Tests/Mapping/ObjectPropertyTest.cs:
```diff
@@ -443,4 +443,27 @@ public void ShouldIgnoreStaticConstructorAndDiagnostic()
+            """
+            [MapProperty("StringValue", "NestedValue.StringValue")]
+            partial B Map(A src);
+""",
```
Weird indention

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/ObjectMemberMappingBodyBuilder.cs:
```diff
@@ -135,6 +135,27 @@ PropertyMappingConfiguration config
+                    member.Name,
+                    member.Type
+                );
+                return true;
```
As discussed in https://github.com/riok/mapperly/issues/516#issuecomment-1655677725 it would make sense to return false here and not generate this mapping at all, wouldn't it?

TimothyMakkison commented:
Sorry, I though I had fixed all of this a week ago. Turns out I forgot to push it 😅

latonz commented:
@TimothyMakkison ah yep on a second thought I think that makes sense, that only the second last is checked.

latonz commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -410,4 +410,13 @@ internal static class DiagnosticDescriptors
+
+    public static readonly DiagnosticDescriptor CannotMapToTemporarySourceMember = new DiagnosticDescriptor(
+        "RMG047",
+        "Cannot map to member path due to modifying a temporary value",
```
I think I'd reference CS1612...

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/ObjectMemberMappingBodyBuilder.cs:
```diff
@@ -132,6 +132,28 @@ PropertyMappingConfiguration config
         }
 
+        // cannot assign to intermediate value type
+        // invalid mapping when second to last member is a property that returns a value type
```
Probably reference CS1612 in the comment.

TimothyMakkison commented:
Added CS1612 to error message.

Modified the check to iterate backwards, checking for value type properties. Checking the penultimate member would work if field types didn't exist

TimothyMakkison commented:
Analyzer doesn't like the length of `ValidateMappingSpecification`. I've extracted the struct method but it's still too long.
Might be a bit controversial but I've temporarily disabled the warning, I'd argue a a long method with simple checks is easier to understand than one method calling 6 simple methods. WDYT?

latonz commented:
IMO its okay here to ignore the warning, could you change it to suppressmessage attribute? Often the readability is improved that when reading a method and submethods are called instead of inlined code, code which I'm not interested in reading can be skipped easier.

TimothyMakkison commented:
> IMO its okay here to ignore the warning, could you change it to suppressmessage attribute?

Added attribute
> Often the readability is improved that when reading a method and submethods are called instead of inlined code, code which I'm not interested in reading can be skipped easier.

Should I extract all the methods?

latonz commented:
Nope, should be fine like this, just my general opinion on the topic…
